### PR TITLE
Rework the API

### PR DESCRIPTION
**Pull request by greg0ire**
I am targeting this branch, because this is a major BC break.

## Changelog

```markdown
### Changed
- The public API has been closed. Use decoration instead of inheritance.
- php 7.1 type hints have been added when possible
```

### CODE REVIEW

File: src/Model/ManagerInterface.php [hunk 1]
```diff
@@ -73,27 +73,23 @@ public function create();
...
-     *
-     * @return string
      */
-    public function getTableName();
+    public function getTableName(): string;
 
     /**
      * Get the DB driver connection.
-     *
-     * @return Connection
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply src/Model/ManagerInterface.php]: Not really, I think it's not the same for MongoDB or PHPCR

File: src/Test/AbstractWidgetTestCase.php [hunk 2]
```diff
@@ -78,18 +76,12 @@ protected function setUp(): void
         }
     }
 
-    /**
-     * @return TwigRendererInterface
-     */
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply src/Test/AbstractWidgetTestCase.php]: Nope, it seems to change

File: src/Test/AbstractWidgetTestCase.php [hunk 3]
```diff
@@ -78,18 +76,12 @@ protected function setUp(): void
...
     final public function getRenderer()
     {
         return $this->renderer;
     }
 
-    /**
-     * @return \Twig_Environment
-     */
-    protected function getEnvironment()
+    final protected function getEnvironment(): \Twig_Environment
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply src/Test/AbstractWidgetTestCase.php]: It should, but that would be for another PR

File: src/Validator/ErrorElement.php [hunk 4]
```diff
@@ -76,7 +76,7 @@ public function __construct(
         $subject,
         ConstraintValidatorFactoryInterface $constraintValidatorFactory,
         $context,
-        $group
+        ?string $group
     ) {
         if (!($context instanceof LegacyExecutionContextInterface) && !($context instanceof ExecutionContextInterface)) {
             throw new \InvalidArgumentException(sprintf('Argument 3 passed to %s::__construct() must be an instance of Symfony\Component\Validator\ExecutionContextInterface or Symfony\Component\Validator\Context\ExecutionContextInterface.', \get_class($this)));
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply src/Validator/ErrorElement.php]: We should make a PR to remove deprecations

File: src/Validator/Constraints/InlineConstraint.php [hunk 5]
```diff
@@ -107,9 +104,6 @@ public function getRequiredOptions()
         ];
     }
 
-    /**
-     * @return string
-     */
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply src/Validator/Constraints/InlineConstraint.php]: I will resurrect it to `string|callable`

File: src/Validator/ErrorElement.php [hunk 6]
```diff
@@ -172,12 +153,9 @@ public function getSubject()
 
     /**
      * @param string|array $message
-     * @param array        $parameters
      * @param null         $value
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply src/Validator/ErrorElement.php]: probably not.

**Comment by kunicmarko20** [comment]
Travis wont start :thinking:

**Comment by greg0ire** [comment]
Tried force pushing, nothing, and https://www.traviscistatus.com/ is not entirely green…

**Comment by greg0ire** [comment]
There's a build, it's just that Github isn't aware of that: https://travis-ci.org/sonata-project/SonataCoreBundle

**Comment by kunicmarko20** [comment]
I restarted the old one, I thought it would start it here but no

**Comment by kunicmarko20** [comment]
Travis started 🎉🎉🎉🎉

**Comment by OskarStark** [comment]
👍🏻

**Pull request merged**